# Weather Data Pipeline

## Action Making Steps:

##### 1: Create Directory Structure
I opened the codespace of my repository called computer infrastructure and with command ``mkdir data`` I created a directory called data at the root of the repository. Inside of directory data, I created 2 subdirectories called timestamps and weather with commands ``mkdir timestamps`` and ``mkdir weather``

#### 2: Timestamps Tests
In the data/timestamps directory, I created file now.txt with command ``date >> now.txt``. I repeated this step few times. Then I checked content of file now.txt with command ``more now.txt`.
With command ``date +"%Y%m%d_%H%M%S" >> formatted.txt``, I created file formatted.txt containing timestamp in required format of YYYYmmdd_HHMMSS.
With command ``touch `date +"%Y%m%d_%H%M%S".txt` ``, I created a file called 20241108_120500.txt.

##### 3: Download Today's Weather Data
I used command ``wget -O weather.json https://prodapi.metweb.ie/observations/athenry/today`` to download weather data for the Athenry weather station from Met Eireann into file called weather.json.
Then with command ``wget -O data/weather/ `date +"%Y%m%d_%H%M%S_Athenry.json"` https://prodapi.metweb.ie/observations/athenry/today`` I downloaded current weather data into a file with timestamp as the name.

##### 7: Script
I wrote script called weather.sh, at the root of my repository:
```
#! /bin/bash

wget -O data/weather/ `date +"%Y%m%d_%H%M%S_Athenry.json"` https://prodapi.metweb.ie/observations/athenry/today
```
This script automates process from step 3. I made the script executable with command ``chmod u+x ./weather.sh.``

## Commands used:
- mkdir - creating directories, if they do not already exist.
- touch - updating access and modification times of each file to the current times. It can be used for creating new file as touch name_of_file.suffix
- date - printing or setting the date and time
- wget - non-interactive network downloading
- more - for view of chosen text file in the command line
- man - displaying user manual for any command that can be run in the command line
- chmod - changing file mode
- pwd - print working directiory - for writting full path of current working directory
- ls - to list the directory content
- cd - changing directory

## Data Analyses for November 9, 2024

Examination and summarisation of data:

In [ ]:
# Import modules
import json
import pandas as pd

In [2]:
# downloading dataset weather.json
data=pd.read_json("data/weather/weather.json")
data

name  temperature symbol weatherDescription               text  \
0   Athenry           12    04n             Cloudy           "Cloudy"   
1   Athenry           12    04n             Cloudy           "Cloudy"   
2   Athenry           12    15n         Fog / Mist             "Mist"   
3   Athenry           12    15n         Fog / Mist             "Mist"   
4   Athenry           12    15n         Fog / Mist             "Mist"   
5   Athenry           12    15n         Fog / Mist             "Mist"   
6   Athenry           12    15n         Fog / Mist             "Mist"   
7   Athenry           12    04n             Cloudy           "Cloudy"   
8   Athenry           12    04d             Cloudy           "Cloudy"   
9   Athenry           12    04d             Cloudy           "Cloudy"   
10  Athenry           12    04d             Cloudy           "Cloudy"   
11  Athenry           13    04d             Cloudy           "Cloudy"   
12  Athenry           13    04d             Cloudy           "Cloudy"   
13  Athenry           14    04d             Cloudy           "Cloudy"   
14  Athenry           15    04d             Cloudy           "Cloudy"   
15  Athenry           14    04d             Cloudy           "Cloudy"   
16  Athenry           14    04d             Cloudy           "Cloudy"   
17  Athenry           14    04n             Cloudy           "Cloudy"   
18  Athenry           14    04n             Cloudy           "Cloudy"   
19  Athenry           13    04n             Cloudy           "Cloudy"   
20  Athenry           13    46n        Light rain    "Light Drizzle "   
21  Athenry           13    46n        Light rain    "Heavy Drizzle "   
22  Athenry           12    46n        Light rain   "Recent Drizzle "   
23  Athenry           12    04n             Cloudy           "Cloudy"   

    windSpeed windGust cardinalWindDirection  windDirection  humidity  \
0          13        -                     E             90        92   
1          15        -                     E             90        93   
2          11        -                     E             90        95   
3          11        -                     E             90        95   
4          11        -                    SE            135        93   
5           9        -                    SE            135        92   
6          11        -                    SE            135        92   
7           9        -                    SE            135        92   
8          11        -                    SE            135        93   
9          11        -                    SE            135        92   
10         11        -                    SE            135        91   
11          7        -                    SE            135        91   
12          9        -                    SE            135        89   
13         11        -                    SE            135        88   
14         11        -                    SE            135        83   
15          6        -                    SE            135        85   
16         15        -                    SE            135        84   
17          7        -                     S            180        85   
18         13        -                     S            180        87   
19          9        -                    SE            135        89   
20          9        -                     S            180        91   
21          7        -                     S            180        96   
22          6        -                     S            180        95   
23          4        -                     S            180        95   

    rainfall  pressure   dayName       date reportTime  
0       0.00      1020  Saturday 2024-09-11      00:00  
1       0.00      1020  Saturday 2024-09-11      01:00  
2       0.00      1020  Saturday 2024-09-11      02:00  
3       0.00      1020  Saturday 2024-09-11      03:00  
4       0.00      1020  Saturday 2024-09-11      04:00  
5       0.00 

In [3]:
# Information about dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   name                   24 non-null     object        
 1   temperature            24 non-null     int64         
 2   symbol                 24 non-null     object        
 3   weatherDescription     24 non-null     object        
 4   text                   24 non-null     object        
 5   windSpeed              24 non-null     int64         
 6   windGust               24 non-null     object        
 7   cardinalWindDirection  24 non-null     object        
 8   windDirection          24 non-null     int64         
 9   humidity               24 non-null     int64         
 10  rainfall               24 non-null     float64       
 11  pressure               24 non-null     int64         
 12  dayName                24 non-null     object        
 13  date   

This dataset contains a list of observations for every hour for November, 9, 2024, for synoptic station in Athenry, Co Galway. 14 Columns: Name, dry bulb temperature in whole degrees, symbol, weather descriptions, windspeed (kt), windgust, cardinal wind direction, relative humidity (%), rainfall (mm), msl Pressure (mbar), the day of the week, date and time of observation. Wind directions here means direction of the wind in 360 ° where 180 ° means south direction, 90 ° means eastern direction, 0 ° means northern direction and 270 ° means western direction. 

https://data.gov.ie/dataset/todays-weather-athenry/resource/2946c036-26fc-4364-9535-f794e936930b

https://dev.qweather.com/en/docs/resource/wind-info

In [17]:
data[['temperature','humidity', 'windSpeed','rainfall', 'pressure']].describe()

temperature   humidity  windSpeed   rainfall     pressure
count    24.000000  24.000000  24.000000  24.000000    24.000000
mean     12.750000  90.750000   9.875000   0.063333  1020.083333
std       0.944089   3.744561   2.771164   0.285439     0.583592
min      12.000000  83.000000   4.000000   0.000000  1019.000000
25%      12.000000  88.750000   8.500000   0.000000  1020.000000
50%      12.000000  92.000000  11.000000   0.000000  1020.000000
75%      13.250000  93.000000  11.000000   0.000000  1020.000000
max      15.000000  96.000000  15.000000   1.400000  1021.000000

For November 9, 2024, average temperature was 12.75 °C, the lowest temperature 12 °C, the highest temperature 15 °C. Average windspeed was 9.87 knots (18.28 km/h), highest speed 15.00 knots (27.78 km/h), the lowest speed 4 knots (7.4 km/h). Average humidity was 90.75 %, the highest humidity 96 %, the lowest 83 %. Average atmospheric pressure was changing only a little during the day with the average of 1020 MPa. Rainfall during the day was from 0.00 mm to 1.4 mm per hour. 

https://www.inchcalculator.com/convert/knot-to-kilometer-per-hour.

In [18]:
data['cardinalWindDirection'].value_counts()

cardinalWindDirection
SE    14
S      6
E      4
Name: count, dtype: int64

Prevailing direction of wind was southeastern.

In [20]:
data['weatherDescription'].value_counts()

weatherDescription
Cloudy         16
Fog / Mist      5
Light rain      3
Name: count, dtype: int64

Weather in November, 9, 2024 was predominantly cloudy with occassions of fog and light rain.

***
### End